probe/awsecs/client.go

```diff
@@ -1,102 +1,212 @@
-			// describe each page of 10 (the max for one describe command) concurrently
-			group.Add(1)
-			serviceArns := page.ServiceArns
-			go func() {
-				defer group.Done()
-
-				resp, err := c.client.DescribeServices(&ecs.DescribeServicesInput{
-					Cluster:  &c.cluster,
-					Services: serviceArns,
-				})
-				if err != nil {
-					log.Warnf("Error describing some ECS services, ECS service report may be incomplete: %v", err)
-					return
-				}
+func newECSService(service *ecs.Service) EcsService {
+	deploymentIDs := make([]string, len(service.Deployments))
+	for i, deployment := range service.Deployments {
+		deploymentIDs[i] = *deployment.Id
+	}
+	return EcsService{
+		ServiceName:       *service.ServiceName,
+		DeploymentIDs:     deploymentIDs,
+		DesiredCount:      *service.DesiredCount,
+		PendingCount:      *service.PendingCount,
+		RunningCount:      *service.RunningCount,
+		TaskDefinitionARN: *service.TaskDefinition,
+	}
+}
 
-				for _, failure := range resp.Failures {
-					log.Warnf("Failed to describe ECS service %s, ECS service report may be incomplete: %s", *failure.Arn, *failure.Reason)
+// Returns a channel from which service ARNs can be read.
+// Cannot fail as it will attempt to deliver partial results, though that may end up being no results.
+func (c ecsClientImpl) listServices() <-chan string {
+	log.Debugf("Listing ECS services")
+	results := make(chan string)
+	go func() {
+		count := 0
+		err := c.client.ListServicesPages(
+			&ecs.ListServicesInput{Cluster: &c.cluster},
+			func(page *ecs.ListServicesOutput, lastPage bool) bool {
+				for _, arn := range page.ServiceArns {
+					count++
+					results <- *arn
 				}
+				return true
+			},
+		)
+		if err != nil {
+			log.Warnf("Error listing ECS services, ECS service report may be incomplete: %v", err)
+		}
+		log.Debugf("Listed %d services", count)
+		close(results)
+	}()
+	return results
+}
 
-				lock.Lock()
-				for _, service := range resp.Services {
-					results[*service.ServiceName] = service
-				}
-				lock.Unlock()
-			}()
-			return true
-		},
-	)
-	group.Wait()
+// Returns (input, done) channels. Service ARNs given to input are batched and details are fetched,
+// with full EcsService objects being put into the cache. Closes done when finished.
+func (c ecsClientImpl) describeServices() (chan<- string, <-chan struct{}) {
+	input := make(chan string)
+	done := make(chan struct{})
+
+	log.Debugf("Describing ECS services")
 
+	go func() {
+		const maxServices = 10 // How many services we can put in one Describe command
+		group := sync.WaitGroup{}
+
+		// count and calls is just for logging
+		count := 0
+		calls := 0
+
+		batch := make([]string, 0, maxServices)
+		for arn := range input {
+			batch = append(batch, arn)
+			if len(batch) == maxServices {
+				group.Add(1)
+				go func(arns []string) {
+					defer group.Done()
+					c.describeServicesBatch(arns)
+				}(batch)
+				count += len(batch)
+				calls++
+				batch = make([]string, 0, maxServices)
```

Comment:
This will cause the next batch to re-use the same in-memory array as the previous batch. Since we haven't made a copy of the underlying array when passing batch to `describeServicesBatch`, only of the slice (which continues pointing to the same underlying array), this causes subsequent appends to batch to modify entries in the previous batch.

Simplified example: https://play.golang.org/p/FfQf7pMqDK